apps/total-typescript/src/server/routers/exercise.ts

```diff
@@ -2,17 +2,41 @@ import {createRouter} from 'server/createRouter'
 import {z} from 'zod'
 import {sanityClient} from 'utils/sanity-client'
 import groq from 'groq'
-import {Exercise, ExerciseSchema} from '../../lib/exercises'
+import {
+  Exercise,
+  ExerciseMedia,
+  ExerciseMediaSchema,
+  ExerciseSchema,
+  getExerciseMedia,
+} from '../../lib/exercises'
 
-export const exerciseRouter = createRouter().query('bySlug', {
-  input: z.object({
-    slug: z.string(),
-  }),
-  async resolve({ctx, input}) {
-    const exercise: Exercise = await sanityClient.fetch(
-      groq`*[_type == "exercise" && slug.current == $slug][0]`,
-      {slug: input.slug},
-    )
-    return ExerciseSchema.parse(exercise)
-  },
-})
+import {createAppAbility} from '../../ability/ability'
+import {ContentRulesSchema, getContentRules} from '../get-content-rules'
+
+export const exerciseRouter = createRouter()
+  .query('bySlug', {
+    input: z.object({
+      slug: z.string(),
+    }),
+    async resolve({ctx, input}) {
+      const exercise: Exercise = await sanityClient.fetch(
+        groq`*[_type == "exercise" && slug.current == $slug][0]`,
+        {slug: input.slug},
+      )
+      return ExerciseSchema.parse(exercise)
+    },
+  })
+  .query('media', {
+    input: ContentRulesSchema,
+    async resolve({ctx, input}) {
+      const rules = await getContentRules(ctx.req, input)
+      const ability = createAppAbility(rules)
+
+      if (input.lessonSlug && ability.can('view', 'Content')) {
+        const exerciseMedia: ExerciseMedia = await getExerciseMedia(
+          input.lessonSlug,
+        )
+        return ExerciseMediaSchema.parse(exerciseMedia)
+      }
```

Comment:
this is a great solution to getting media URLs behind server-side ability checks 💯